lib/Sema/CSDiag.cpp

```diff
@@ -6243,8 +6190,24 @@ bool FailureDiagnosis::visitAssignExpr(AssignExpr *assignExpr) {
       if (diagnoseSubscriptErrors(subscriptExpr, /* inAssignmentDestination = */ true))
         return true;
     }
-    if (CS.diagnoseAssignmentFailure(destExpr, destType, assignExpr->getLoc()))
+    // Member ref assignment errors detected elsewhere, so not an assignment issue if found here.
+    // The remaining exception involves mutable pointer conversions which aren't always caught elsewhere.
+    PointerTypeKind ptk;
+    if (!isa<MemberRefExpr>(destExpr) || CS.getType(destExpr)
+                                             ->lookThroughAllOptionalTypes()
+                                             ->getAnyPointerElementType(ptk)) {
```

Comment:
I think it's reasonable to get this whole `if` into `findDeclDiagonstic` when https://github.com/apple/swift/pull/18995 lands